rfd/0123-tls-routing-behind-layer7-lb.md

```diff
@@ -299,15 +321,58 @@ for different clusters.
 UX is the same as when [TLS Routing is
 enabled](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md#local-teleport-proxy).
 
+#### 5 - Client source IPs
+
+A new setting is added to the Proxy service for web APIs to take client IPs from the "X-Forwarded-For" headers:
+```
+proxy_service:
+  # Enables the Proxy service to take client source IPs from the
+  # "X-Forwarded-For" headers for web APIs recevied from layer 7 load balancers
+  # or reverse proxies.
+  # 
+  # IMPORTANT: please make sure the service is behind a layer 7 load balancer or
+  # reverse proxy that sets or appends client IPs in the "X-Forwarded-For"
+  # headers.
+  #
+  # default: false
+  use_xff_header: true
+```
+
+Note that this setting is NOT mutually exclusive with the `proxy_protocal` setting. For example, in a separate-port
+setup, web APIs including the ALPN connection upgrade endpoint (e.g. called by `tsh proxy db`) can take client source
+IPs from the XFF headers set by a layer 7 load balancer, whereas SSH, reverse tunnels and separate DB ports can take
+client source IPs from the PROXY headers set by a layer 4 load balancer.
+
 ### Security
 
+#### 1 - Connection upgrade
+
 When upgrading the connection, the Teleport client verifies the load balancer's TLS cert using SystemCertPool. And as
 mentioned early, at the TLS Routing request, the Teleport client will be configured with a Teleport CA for verifying the
 Proxy server.
 
 There is no authentication at the connection upgrade request. Authentication is deferred to the TLS Routing request so
 authentication remains the same as if there is no connection upgrade.
 
+#### 2 - Client source IPs
+
+The premise of the IP Pinning feature is to protect against hackers using comproised credentials from different
+locations.
+
+A hacker can easily modify a Teleport client with a fake IP in the "X-Forwarded-For" (XFF) if the Proxy service trusts
+this value without conditions. Remember the ALPN connection upgrade endpoint can embed all Teleport protocols in TLS
+routing requests, and the endpoint is publicly available regardless of whether the Proxy is behind a load balancer or
+not.
+
+To minimize this risk, the web APIs will only take client source IPs from XFF headers when
+`proxy_service.use_xff_header` is explicitly set to true. Users have the responsibility to ensure that the Proxy
+service is indeed behind a layer 7 load balancer AND the load balancer is configured to set correct values in the XFF
+headers.
+
+Further more, requests with multiple XFF values will be rejected as it is difficult to know which is the truth. And in
+cases multiple IPs are present in the header, the last IP is taken as only the last one can be trusted as the load
+balancer sets it before delivering the request to the Proxy.
```

Comment:
The two sentences seem to contradict each other.

Since we deny the request if `len(req.Header.Values("XFF"))>1`, there is no reason to pick the last because we will only pick the single element because `len(req.Header.Values("XFF"))==1`.

Don't forget to use `header.Values` because Get will only pick the first entry

Reply:
@tigrato Sorry maybe I was confused with the previous comments on this. I thought you were referring to the case where `len(req.Header.Values) > 0` and that should be rejected. When `len(req.Header.Values) == 1`, `req.Header.Values[0]` can still hold multiple addresses. I was thinking get the last of `req.Header.Values[0]`. 

are you suggesting we reject if `len(strings.Split(strings.Join(req.Header.Values("XFF"), ","), ",")) > 1` ?

Reply:
Yes.
L7 proxies append the seen IP address to a single XFF entry joined with commas.
HTTP allows sending multiple times the same header but in this case, the output is always just a single header entry with a single value. The value can have multiple 

For me, I think we should just accept if the XFF only contains a single IP address. If we have more we either:
- use the last but we can use the chained L7 IP and everything will work but we never really implement IP pinning for them.
- having a user sending spam headers to test behaviors.

From a security perspective, we should only accept the XFF if the headers only have one entry and if that entry only contains a single IP address. Users can disable XFF append and let it through but it's the user's responsibility to have it working.

Maybe I am enforcing too much but advertising the IP pinning and having no way to confirm if it's working (with multiple entries or single entry with multiple IPS) is a big problem. 

We should keep it simple and ensure it works.

Reply:
@tigrato's recommendation here would dramatically reduce the risks for customers who have incorrect proxy configurations we discussed here: https://github.com/gravitational/teleport/pull/26620#discussion_r1205876661

If palatable, I would fully support making this change.  The obvious issue is that customers could not have multiple levels of proxies (which does occur in some network configurations).

Reply:
@tigrato @jentfoo awesome! I'd love to simplify to allow a single ip only. I just got confused with the terminologies. Will update.

As discussed early, a single L7 LB should be the common setup. If a customer has problem getting a chain setup working, they can make a feature request to support that (e.g. [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) suggests `Trusted proxy count` and `Trusted proxy list`) but we can leave it out the scope for now.

Reply:
Thank you @greedy52 !

Reply:
Yep, I agree with @tigrato, only accepting one value seems like a good solution to start with and it's consistent with our PROXY protocol handling - we only accept one external PROXY header and deny connection otherwise. I also mentioned creating trusted proxies config in the IP pinning RFD as a possible future improvement, it applies here as well (it was not implemented yet).